DDG4/edm4hep/EDM4hepFileReader.cpp

```diff
@@ -0,0 +1,328 @@
+    }
+    for(auto const& key: source.template getKeys<float>()) {
+      m_fltValues[key] = source.template get<std::vector<float>>(key).value();
+    }
+    for(auto const& key: source.template getKeys<double>()) {
+      m_dblValues[key] = source.template get<std::vector<double>>(key).value();
+    }
+    for(auto const& key: source.template getKeys<std::string>()) {
+      m_strValues[key] = source.template get<std::vector<std::string>>(key).value();
+    }
+  }
+
+  /// Class to read EDM4hep files
+  /**
+   *  \version 1.0
+   *  \ingroup DD4HEP_SIMULATION
+   */
+  class EDM4hepFileReader : public Geant4EventReader {
+  protected:
+    /// Reference to reader object
+    podio::Reader m_reader;
+    /// Name of the MCParticle collection to read
+    std::string m_collectionName;
+    /// Name of the EventHeader collection to read
+    std::string m_eventHeaderCollectionName;
+
+  public:
+    /// Initializing constructor
+    EDM4hepFileReader(const std::string& nam);
+
+    /// Read parameters set for this action
+    virtual EventReaderStatus setParameters(std::map< std::string, std::string >& parameters);
+
+    /// Read an event and fill a vector of MCParticles.
+    virtual EventReaderStatus readParticles(int event_number, Vertices& vertices, std::vector<Particle*>& particles);
+    /// Call to register the run parameters
+    void registerRunParameters();
+
+  };
+
+  /// Initializing constructor
+  dd4hep::sim::EDM4hepFileReader::EDM4hepFileReader(const std::string& nam)
+    : Geant4EventReader(nam)
+    , m_reader(podio::makeReader(nam))
+    , m_collectionName("MCParticles")
+    , m_eventHeaderCollectionName("EventHeader")
+  {
+    printout(INFO,"EDM4hepFileReader","Created file reader. Try to open input %s",nam.c_str());
+    m_directAccess = true;
+  }
+
+  void EDM4hepFileReader::registerRunParameters() {
+    try {
+      auto *parameters = new RunParameters();
+      try {
+        podio::Frame runFrame = m_reader.readFrame("runs", 0);
+        parameters->ingestParameters(runFrame.getParameters());
+      } catch (std::runtime_error& e) {
+        // we ignore if we do not have runs information
+      } catch(std::invalid_argument&) {
+        // we ignore if we do not have runs information
+      }
+      context()->run().addExtension<RunParameters>(parameters);
+
+      auto *fileParameters = new FileParameters();
+      try {
+        podio::Frame metaFrame = m_reader.readFrame("metadata", 0);
+        fileParameters->ingestParameters(metaFrame.getParameters());
+      } catch (std::runtime_error& e) {
+        // we ignore if we do not have metadata information
+      } catch(std::invalid_argument&) {
+        // we ignore if we do not have metadata information
+      }
+      context()->run().addExtension<FileParameters>(fileParameters);
+    } catch(std::exception &e) {
+      printout(ERROR,"EDM4hepFileReader::registerRunParameters","Failed to register run parameters: %s", e.what());
```

Comment:
```suggestion
      printout(ERROR,"EDM4hepFileReader::registerRunParameters","Failed to register run or file parameters: %s", e.what());
```
Unless *run* is referring to the DD4hep run and not the EDM4hep / podio run(?).

Reply:
I split the catch between file and run parameters and updated the error message.